File: src/main/kotlin/io/vinicius/rmd/helper/Telemetry.kt
package io.vinicius.rmd.helper

import com.mixpanel.mixpanelapi.ClientDelivery
import com.mixpanel.mixpanelapi.MessageBuilder
import com.mixpanel.mixpanelapi.MixpanelAPI
import org.json.JSONObject
import java.util.UUID

object Telemetry {
    private val mixpanel = MixpanelAPI()
    private val messageBuilder = MessageBuilder("5be99deb1f8e8bdd9cca5e8e0a1a15a8")
    private val trackingId = UUID.randomUUID().toString()

    fun trackDownloadStart(
        user: String,
        limit: Int,
        parallel: Int,
        convertImages: Boolean,
        convertVideos: Boolean
    ) {
        val props = JSONObject(mapOf(
            "user" to user,
            "limit" to limit,
            "parallel" to parallel,
            "convertImages" to convertImages,
            "convertVideos" to convertVideos
        ))

        val event = messageBuilder.event(trackingId, "Download Start", props)
        val delivery = ClientDelivery().apply { addMessage(event) }
        mixpanel.deliver(delivery)
    }

    fun trackDownloadEnd(
        user: String,
        postsFound: Int,
        failedDownloads: Int,
        duplicated: Int
    ) {
        val props = JSONObject(mapOf(
            "user" to user,
            "postsFound" to postsFound,
            "failedDownloads" to failedDownloads,
            "duplicated" to duplicated
        ))

        val event = messageBuilder.event(trackingId, "Download End", props)
        val delivery = ClientDelivery().apply { addMessage(event) }
        mixpanel.deliver(delivery)
    }
}

File: telemetry.go
package main

import (
	"context"
	"github.com/google/uuid"
	"github.com/mixpanel/mixpanel-go"
)

var mp = mixpanel.NewApiClient("5be99deb1f8e8bdd9cca5e8e0a1a15a8")
var distinctId = uuid.New().String()

func TrackDownloadStart(
	version string,
	source string,
	name string,
	parallel int,
	limit int,
	convertImages bool,
	convertVideos bool,
) {
	ctx := context.Background()

	_ = mp.Track(ctx, []*mixpanel.Event{
		mp.NewEvent("Reddit Download Start", distinctId, map[string]any{
			"version":       version,
			"source":        source,
			"name":          name,
			"parallel":      parallel,
			"limit":         limit,
			"convertImages": convertImages,
			"convertVideos": convertVideos,
		}),
	})
}

func TrackDownloadEnd(
	version string,
	source string,
	name string,
	mediaFound int,
	failedDownloads int,
	duplicated int,
) {
	ctx := context.Background()

	_ = mp.Track(ctx, []*mixpanel.Event{
		mp.NewEvent("Reddit Download End", distinctId, map[string]any{
			"version":         version,
			"source":          source,
			"name":            name,
			"mediaFound":      mediaFound,
			"failedDownloads": failedDownloads,
			"duplicated":      duplicated,
		}),
	})
}
